File: tests/php/MountTest.php
<?php

declare(strict_types=1);

namespace AgentHarness\Tests;

use AgentHarness\BuiltinFilesystemDriver;
use AgentHarness\HasShell;
use AgentHarness\LocalFolderSource;
use AgentHarness\Mountable;
use AgentHarness\MountingFilesystemDriver;
use AgentHarness\MountSource;
use AgentHarness\MountStat;
use AgentHarness\UsesTools;
use PHPUnit\Framework\TestCase;

/** In-memory read-only source. `files` maps relative subpaths -> content. */
class FakeSource implements MountSource
{
    /** @var array<string, string> */
    private array $files = [];

    /** @param array<string, string> $files */
    public function __construct(array $files)
    {
        foreach ($files as $k => $v) {
            $this->files[trim($k, '/')] = $v;
        }
    }

    /** @return array<string, bool> */
    private function dirs(): array
    {
        $dirs = ['' => true];
        foreach (array_keys($this->files) as $p) {
            $parts = explode('/', $p);
            for ($i = 1; $i < count($parts); $i++) {
                $dirs[implode('/', array_slice($parts, 0, $i))] = true;
            }
        }
        return $dirs;
    }

    public function stat(string $subpath): ?MountStat
    {
        $subpath = trim($subpath, '/');
        if (isset($this->files[$subpath])) {
            return new MountStat(isDir: false, size: strlen($this->files[$subpath]));
        }
        if (isset($this->dirs()[$subpath])) {
            return new MountStat(isDir: true);
        }
        return null;
    }

    /** @return list<string> */
    public function list(string $subpath): array
    {
        $subpath = trim($subpath, '/');
        $prefix = $subpath === '' ? '' : $subpath . '/';
        $names = [];
        foreach (array_keys($this->files) as $p) {
            if ($p !== $subpath && str_starts_with($p, $prefix)) {
                $rest = substr($p, strlen($prefix));
                $names[explode('/', $rest)[0]] = true;
            }
        }
        $out = array_keys($names);
        sort($out);
        return $out;
    }

    public function read(string $subpath): string
    {
        return $this->files[trim($subpath, '/')];
    }
}

class MountAgent
{
    use HasShell;
    use UsesTools;
}

class MountTest extends TestCase
{
    private function inner(): BuiltinFilesystemDriver
    {
        return new BuiltinFilesystemDriver();
    }

    private function mounting(): MountingFilesystemDriver
    {
        return new MountingFilesystemDriver($this->inner());
    }

    public function testEmptyTablePassthrough(): void
    {
        $inner = $this->inner();
        $inner->write('/a.txt', 'hi');
        $fs = new MountingFilesystemDriver($inner);
        $this->assertSame('hi', $fs->read('/a.txt'));
        $this->assertSame([], $fs->mounts());
    }

    public function testLongestPrefixRouting(): void
    {
        $fs = $this->mounting();
        $fs->mount('/a', new FakeSource(['x.txt' => 'outer']));
        $fs->mount('/a/b', new FakeSource(['y.txt' => 'inner']));
        $this->assertSame('inner', $fs->read('/a/b/y.txt'));
        $this->assertSame('outer', $fs->read('/a/x.txt'));
    }

    public function testReadFromSource(): void
    {
        $fs = $this->mounting();
        $fs->mount('/work', new FakeSource(['hello.txt' => 'world']));
        $this->assertSame('world', $fs->read('/work/hello.txt'));
        $this->assertSame('world', $fs->readText('/work/hello.txt'));
    }

    public function testExistsMountPointAndAncestor(): void
    {
        $fs = $this->mounting();
        $fs->mount('/deep/mnt', new FakeSource(['f.txt' => 'x']));
        $this->assertTrue($fs->exists('/deep/mnt'));
        $this->assertTrue($fs->exists('/deep'));
        $this->assertTrue($fs->exists('/deep/mnt/f.txt'));
        $this->assertFalse($fs->exists('/deep/mnt/nope.txt'));
        $this->assertTrue($fs->isDir('/deep'));
        $this->assertTrue($fs->isDir('/deep/mnt'));
        $this->assertFalse($fs->isDir('/deep/mnt/f.txt'));
    }

    public function testListdirMergeDedupSorted(): void
    {
        $inner = $this->inner();
        $inner->write('/work/local.txt', 'L');
        $inner->write('/work/shared.txt', 'edited');
        $fs = new MountingFilesystemDriver($inner);
        $fs->mount('/work', new FakeSource(['shared.txt' => 'orig', 'sub/deep.txt' => 'd']));
        $this->assertSame(['local.txt', 'shared.txt', 'sub'], $fs->listdir('/work'));
    }

    public function testListdirRootShowsMountPoint(): void
    {
        $fs = $this->mounting();
        $fs->mount('/repo', new FakeSource(['a.txt' => 'x']));
        $this->assertContains('repo', $fs->listdir('/'));
    }

    public function testCopyUpShadowing(): void
    {
        $fs = $this->mounting();
        $fs->mount('/work', new FakeSource(['hello.txt' => 'original']));
        $this->assertSame('original', $fs->read('/work/hello.txt'));
        $fs->write('/work/hello.txt', 'edited');
        $this->assertSame('edited', $fs->read('/work/hello.txt'));
        $this->assertSame(1, count(array_filter($fs->listdir('/work'), fn($n) => $n === 'hello.txt')));
    }

    public function testRemoveCopiedUpUnshadows(): void
    {
        $fs = $this->mounting();
        $fs->mount('/work', new FakeSource(['hello.txt' => 'original']));
        $fs->write('/work/hello.txt', 'edited');
        $fs->remove('/work/hello.txt');
        $this->assertSame('original', $fs->read('/work/hello.txt'));
    }

    public function testRemoveSourceOnlyRaises(): void
    {
        $fs = $this->mounting();
        $fs->mount('/work', new FakeSource(['hello.txt' => 'original']));
        $this->expectException(\RuntimeException::class);
        $fs->remove('/work/hello.txt');
    }

    public function testFindMerged(): void
    {
        $inner = $this->inner();
        $inner->write('/work/local.md', 'L');
        $fs = new MountingFilesystemDriver($inner);
        $fs->mount('/work', new FakeSource(['a.md' => 'x', 'sub/b.md' => 'y', 'c.txt' => 'z']));
        $found = $fs->find('/work', '*.md');
        sort($found);
        $this->assertSame(['/work/a.md', '/work/local.md', '/work/sub/b.md'], $found);
    }

    public function testStatSourceFile(): void
    {
        $fs = $this->mounting();
        $fs->mount('/work', new FakeSource(['hello.txt' => 'world']));
        $s = $fs->stat('/work/hello.txt');
        $this->assertSame('file', $s['type']);
        $this->assertSame(5, $s['size']);
        $this->assertSame('directory', $fs->stat('/work')['type']);
    }

    public function testCloneWritableDivergesSourceShared(): void
    {
        $fs = $this->mounting();
        $fs->mount('/work', new FakeSource(['hello.txt' => 'original']));
        $fs->write('/work/local.txt', 'L');
        $clone = $fs->cloneFs();
        $this->assertInstanceOf(MountingFilesystemDriver::class, $clone);
        $clone->write('/work/local.txt', 'changed');
        $this->assertSame('L', $fs->read('/work/local.txt'));
        $this->assertSame('changed', $clone->read('/work/local.txt'));
        $this->assertSame($fs->mounts(), $clone->mounts());
        $this->assertSame('original', $clone->read('/work/hello.txt'));
    }

    public function testUnmount(): void
    {
        $fs = $this->mounting();
        $fs->mount('/work', new FakeSource(['a.txt' => 'x']));
        $this->assertSame(['/work'], $fs->mounts());
        $fs->unmount('/work');
        $this->assertSame([], $fs->mounts());
        $this->assertFalse($fs->exists('/work/a.txt'));
    }

    public function testIsMountable(): void
    {
        $this->assertInstanceOf(Mountable::class, $this->mounting());
    }

    // --- LocalFolderSource ---------------------------------------------------

    private function mkTemp(): string
    {
        $dir = sys_get_temp_dir() . '/mount-' . bin2hex(random_bytes(8));
        mkdir($dir, 0o755, true);
        return $dir;
    }

    public function testLocalFolderSourceNestedFiles(): void
    {
        $root = $this->mkTemp();
        file_put_contents($root . '/a.txt', 'alpha');
        mkdir($root . '/sub');
        file_put_contents($root . '/sub/b.txt', 'beta');

        $src = new LocalFolderSource($root);
        $this->assertTrue($src->stat('')->isDir);
        $this->assertFalse($src->stat('a.txt')->isDir);
        $this->assertSame(5, $src->stat('a.txt')->size);
        $list = $src->list('');
        sort($list);
        $this->assertSame(['a.txt', 'sub'], $list);
        $this->assertSame(['b.txt'], $src->list('sub'));
        $this->assertSame('alpha', $src->read('a.txt'));
        $this->assertSame('beta', $src->read('sub/b.txt'));
        $this->assertNull($src->stat('nope.txt'));
    }

    public function testLocalFolderSourceParentTraversalCannotEscape(): void
    {
        $base = $this->mkTemp();
        $root = $base . '/root';
        mkdir($root);
        file_put_contents($root . '/in.txt', 'inside');
        file_put_contents($base . '/secret.txt', 'secret');

        $src = new LocalFolderSource($root);
        $this->assertNull($src->stat('../secret.txt'));
    }

    public function testLocalFolderSourceSymlinkCannotEscape(): void
    {
        $base = $this->mkTemp();
        $root = $base . '/root';
        mkdir($root);
        $outside = $base . '/outside.txt';
        file_put_contents($outside, 'secret');
        if (!@symlink($outside, $root . '/link.txt')) {
            $this->markTestSkipped('symlinks unsupported');
        }
        $src = new LocalFolderSource($root);
        $this->assertNull($src->stat('link.txt'));
    }

    // --- E2E -----------------------------------------------------------------

    public function testMountVisibleToToolsAndShell(): void
    {
        $root = $this->mkTemp();
        file_put_contents($root . '/hello.txt', 'from-host');
        file_put_contents($root . '/notes.md', '# notes');
        mkdir($root . '/sub');
        file_put_contents($root . '/sub/deep.txt', 'deep-content');

        $agent = new MountAgent();
        $agent->mountSource('/work', new LocalFolderSource($root));

        $this->assertStringContainsString('hello.txt', $agent->execCommand('ls /work')->stdout);
        $this->assertStringContainsString('from-host', $agent->execCommand('cat /work/hello.txt')->stdout);
        $this->assertStringContainsString('notes.md', $agent->execCommand("find /work -name '*.md'")->stdout);
        $this->assertStringContainsString('deep-content', $agent->execCommand('grep -r deep /work')->stdout);

        $fs = $agent->shell()->fs();
        $this->assertSame('deep-content', $fs->readText('/work/sub/deep.txt'));
        $this->assertTrue($fs->exists('/work/notes.md'));
    }

    public function testEditMountedFileCopyUpHostUnchanged(): void
    {
        $root = $this->mkTemp();
        $hostFile = $root . '/hello.txt';
        file_put_contents($hostFile, 'from-host');

        $agent = new MountAgent();
        $agent->mountSource('/work', new LocalFolderSource($root));

        $agent->execCommand('echo edited > /work/hello.txt');
        $this->assertSame('edited', trim($agent->execCommand('cat /work/hello.txt')->stdout));
        $this->assertSame('from-host', file_get_contents($hostFile));
    }

    public function testMountsListing(): void
    {
        $root = $this->mkTemp();
        $agent = new MountAgent();
        $agent->mountSource('/work', new LocalFolderSource($root));
        $this->assertSame(['/work'], $agent->mountSources());
        $agent->unmountSource('/work');
        $this->assertSame([], $agent->mountSources());
    }
}

File: src/go/shell/vfs/mount_test.go
package vfs

import (
	"os"
	"path/filepath"
	"sort"
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// fakeSource is an in-memory read-only MountSource for wrapper unit tests.
type fakeSource struct {
	files map[string][]byte
}

func newFakeSource(files map[string]string) *fakeSource {
	m := make(map[string][]byte, len(files))
	for k, v := range files {
		m[k] = []byte(v)
	}
	return &fakeSource{files: m}
}

func (s *fakeSource) dirs() map[string]struct{} {
	dirs := map[string]struct{}{"": {}}
	for p := range s.files {
		segs := strings.Split(p, "/")
		for i := 1; i < len(segs); i++ {
			dirs[strings.Join(segs[:i], "/")] = struct{}{}
		}
	}
	return dirs
}

func (s *fakeSource) Stat(subpath string) (MountStat, bool) {
	subpath = trimSlashes(subpath)
	if c, ok := s.files[subpath]; ok {
		return MountStat{IsDir: false, Size: int64(len(c))}, true
	}
	if _, ok := s.dirs()[subpath]; ok {
		return MountStat{IsDir: true}, true
	}
	return MountStat{}, false
}

func (s *fakeSource) List(subpath string) ([]string, error) {
	subpath = trimSlashes(subpath)
	prefix := ""
	if subpath != "" {
		prefix = subpath + "/"
	}
	set := map[string]struct{}{}
	for p := range s.files {
		if len(p) > len(prefix) && p[:len(prefix)] == prefix {
			rest := p[len(prefix):]
			set[strings.SplitN(rest, "/", 2)[0]] = struct{}{}
		}
	}
	out := make([]string, 0, len(set))
	for k := range set {
		out = append(out, k)
	}
	sort.Strings(out)
	return out, nil
}

func (s *fakeSource) Read(subpath string) ([]byte, error) {
	return s.files[trimSlashes(subpath)], nil
}

func trimSlashes(p string) string {
	for len(p) > 0 && p[0] == '/' {
		p = p[1:]
	}
	for len(p) > 0 && p[len(p)-1] == '/' {
		p = p[:len(p)-1]
	}
	return p
}

func newMounting() *MountingFilesystemDriver {
	return NewMountingFilesystemDriver(NewBuiltinFilesystemDriver())
}

func TestMountEmptyTablePassthrough(t *testing.T) {
	inner := NewBuiltinFilesystemDriver()
	require.NoError(t, inner.WriteString("/a.txt", "hi"))
	fs := NewMountingFilesystemDriver(inner)
	got, err := fs.ReadString("/a.txt")
	require.NoError(t, err)
	assert.Equal(t, "hi", got)
	assert.Empty(t, fs.Mounts())
}

func TestMountLongestPrefixRouting(t *testing.T) {
	fs := newMounting()
	fs.Mount("/a", newFakeSource(map[string]string{"x.txt": "outer"}))
	fs.Mount("/a/b", newFakeSource(map[string]string{"y.txt": "inner"}))
	got, err := fs.ReadString("/a/b/y.txt")
	require.NoError(t, err)
	assert.Equal(t, "inner", got)
	got, err = fs.ReadString("/a/x.txt")
	require.NoError(t, err)
	assert.Equal(t, "outer", got)
}

func TestMountReadFromSource(t *testing.T) {
	fs := newMounting()
	fs.Mount("/work", newFakeSource(map[string]string{"hello.txt": "world"}))
	got, err := fs.ReadString("/work/hello.txt")
	require.NoError(t, err)
	assert.Equal(t, "world", got)
}

func TestMountExistsMountPointAndAncestor(t *testing.T) {
	fs := newMounting()
	fs.Mount("/deep/mnt", newFakeSource(map[string]string{"f.txt": "x"}))
	assert.True(t, fs.Exists("/deep/mnt"))
	assert.True(t, fs.Exists("/deep"))
	assert.True(t, fs.Exists("/deep/mnt/f.txt"))
	assert.False(t, fs.Exists("/deep/mnt/nope.txt"))
	assert.True(t, fs.IsDir("/deep"))
	assert.True(t, fs.IsDir("/deep/mnt"))
	assert.False(t, fs.IsDir("/deep/mnt/f.txt"))
}

func TestMountListdirMergeDedupSorted(t *testing.T) {
	inner := NewBuiltinFilesystemDriver()
	require.NoError(t, inner.WriteString("/work/local.txt", "L"))
	require.NoError(t, inner.WriteString("/work/shared.txt", "edited"))
	fs := NewMountingFilesystemDriver(inner)
	fs.Mount("/work", newFakeSource(map[string]string{"shared.txt": "orig", "sub/deep.txt": "d"}))
	got, err := fs.Listdir("/work")
	require.NoError(t, err)
	assert.Equal(t, []string{"local.txt", "shared.txt", "sub"}, got)
}

func TestMountListdirRootShowsMountPoint(t *testing.T) {
	fs := newMounting()
	fs.Mount("/repo", newFakeSource(map[string]string{"a.txt": "x"}))
	got, err := fs.Listdir("/")
	require.NoError(t, err)
	assert.Contains(t, got, "repo")
}

func TestMountCopyUpShadowing(t *testing.T) {
	fs := newMounting()
	fs.Mount("/work", newFakeSource(map[string]string{"hello.txt": "original"}))
	got, _ := fs.ReadString("/work/hello.txt")
	assert.Equal(t, "original", got)
	require.NoError(t, fs.WriteString("/work/hello.txt", "edited"))
	got, _ = fs.ReadString("/work/hello.txt")
	assert.Equal(t, "edited", got)
	names, _ := fs.Listdir("/work")
	count := 0
	for _, n := range names {
		if n == "hello.txt" {
			count++
		}
	}
	assert.Equal(t, 1, count)
}

func TestMountRemoveCopiedUpUnshadows(t *testing.T) {
	fs := newMounting()
	fs.Mount("/work", newFakeSource(map[string]string{"hello.txt": "original"}))
	require.NoError(t, fs.WriteString("/work/hello.txt", "edited"))
	require.NoError(t, fs.Remove("/work/hello.txt"))
	got, err := fs.ReadString("/work/hello.txt")
	require.NoError(t, err)
	assert.Equal(t, "original", got)
}

func TestMountRemoveSourceOnlyRaises(t *testing.T) {
	fs := newMounting()
	fs.Mount("/work", newFakeSource(map[string]string{"hello.txt": "original"}))
	err := fs.Remove("/work/hello.txt")
	assert.Error(t, err)
	// source still visible
	got, _ := fs.ReadString("/work/hello.txt")
	assert.Equal(t, "original", got)
}

func TestMountFindMerged(t *testing.T) {
	inner := NewBuiltinFilesystemDriver()
	require.NoError(t, inner.WriteString("/work/local.md", "L"))
	fs := NewMountingFilesystemDriver(inner)
	fs.Mount("/work", newFakeSource(map[string]string{"a.md": "x", "sub/b.md": "y", "c.txt": "z"}))
	got, err := fs.Find("/work", "*.md")
	require.NoError(t, err)
	sort.Strings(got)
	assert.Equal(t, []string{"/work/a.md", "/work/local.md", "/work/sub/b.md"}, got)
}

func TestMountStatSourceFile(t *testing.T) {
	fs := newMounting()
	fs.Mount("/work", newFakeSource(map[string]string{"hello.txt": "world"}))
	fi, err := fs.Stat("/work/hello.txt")
	require.NoError(t, err)
	assert.Equal(t, "file", fi.Type)
	assert.Equal(t, int64(5), fi.Size)
	d, err := fs.Stat("/work")
	require.NoError(t, err)
	assert.Equal(t, "dir", d.Type)
}

func TestMountCloneWritableDivergesSourceShared(t *testing.T) {
	fs := newMounting()
	fs.Mount("/work", newFakeSource(map[string]string{"hello.txt": "original"}))
	require.NoError(t, fs.WriteString("/work/local.txt", "L"))
	clone := fs.Clone().(*MountingFilesystemDriver)
	require.NoError(t, clone.WriteString("/work/local.txt", "changed"))
	got, _ := fs.ReadString("/work/local.txt")
	assert.Equal(t, "L", got)
	got, _ = clone.ReadString("/work/local.txt")
	assert.Equal(t, "changed", got)
	assert.Equal(t, fs.Mounts(), clone.Mounts())
	got, _ = clone.ReadString("/work/hello.txt")
	assert.Equal(t, "original", got)
}

func TestMountUnmount(t *testing.T) {
	fs := newMounting()
	fs.Mount("/work", newFakeSource(map[string]string{"a.txt": "x"}))
	assert.Equal(t, []string{"/work"}, fs.Mounts())
	fs.Unmount("/work")
	assert.Empty(t, fs.Mounts())
	assert.False(t, fs.Exists("/work/a.txt"))
}

func TestMountIsMountable(t *testing.T) {
	var _ Mountable = newMounting()
}

// ---------------------------------------------------------------------------
// LocalFolderSource vs a real temp dir
// ---------------------------------------------------------------------------

func TestLocalFolderSourceNestedFiles(t *testing.T) {
	root := t.TempDir()
	require.NoError(t, os.WriteFile(filepath.Join(root, "a.txt"), []byte("alpha"), 0o644))
	require.NoError(t, os.Mkdir(filepath.Join(root, "sub"), 0o755))
	require.NoError(t, os.WriteFile(filepath.Join(root, "sub", "b.txt"), []byte("beta"), 0o644))

	src := NewLocalFolderSource(root)
	st, ok := src.Stat("")
	require.True(t, ok)
	assert.True(t, st.IsDir)
	st, ok = src.Stat("a.txt")
	require.True(t, ok)
	assert.False(t, st.IsDir)
	assert.Equal(t, int64(5), st.Size)

	names, err := src.List("")
	require.NoError(t, err)
	sort.Strings(names)
	assert.Equal(t, []string{"a.txt", "sub"}, names)
	names, err = src.List("sub")
	require.NoError(t, err)
	assert.Equal(t, []string{"b.txt"}, names)

	data, err := src.Read("a.txt")
	require.NoError(t, err)
	assert.Equal(t, "alpha", string(data))
	data, err = src.Read("sub/b.txt")
	require.NoError(t, err)
	assert.Equal(t, "beta", string(data))

	_, ok = src.Stat("nope.txt")
	assert.False(t, ok)
}

func TestLocalFolderSourceParentTraversalCannotEscape(t *testing.T) {
	base := t.TempDir()
	root := filepath.Join(base, "root")
	require.NoError(t, os.Mkdir(root, 0o755))
	require.NoError(t, os.WriteFile(filepath.Join(root, "in.txt"), []byte("inside"), 0o644))
	require.NoError(t, os.WriteFile(filepath.Join(base, "secret.txt"), []byte("secret"), 0o644))

	src := NewLocalFolderSource(root)
	_, ok := src.Stat("../secret.txt")
	assert.False(t, ok)
}

func TestLocalFolderSourceSymlinkCannotEscape(t *testing.T) {
	base := t.TempDir()
	root := filepath.Join(base, "root")
	require.NoError(t, os.Mkdir(root, 0o755))
	outside := filepath.Join(base, "outside.txt")
	require.NoError(t, os.WriteFile(outside, []byte("secret"), 0o644))
	if err := os.Symlink(outside, filepath.Join(root, "link.txt")); err != nil {
		t.Skip("symlinks unsupported")
	}

	src := NewLocalFolderSource(root)
	_, ok := src.Stat("link.txt")
	assert.False(t, ok)
}
